File: solutions/0116-Populating-Next-Right-Pointers-in-Each-Node/0116.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */


 // 将每两个相邻节点连接起来
 func connectTwoNode(node1 *Node, node2 *Node) {
	 if node1 == nil && node2 == nil {
		 return
	 }
	 node1.Next = node2

	 // 链接相同父节点的两个子节点
	 connectTwoNode(node1.Left, node1.Right)
	 connectTwoNode(node2.Left, node2.Right)
	 // 链接不同父节点的两个子节点
	 connectTwoNode(node1.Right, node2.Left)
 }

 func connect(root *Node) *Node {
	if root == nil {
		return nil
	}

	connectTwoNode(root.Left, root.Right)
	return root
}

File: solutions/0116-Populating-Next-Right-Pointers-in-Each-Node/0116.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None
        
        # 将每两个相邻节点连接起来
        def connectTwoNode(node1: 'Node', node2: 'Node'):
            if not node1 or not node2:
                return
            
            node1.next = node2
            
            # 链接相同父节点的两个子节点
            connectTwoNode(node1.left, node1.right)
            connectTwoNode(node2.left, node2.right)
            # 链接不同父节点的两个子节点
            connectTwoNode(node1.right, node2.left)

        connectTwoNode(root.left, root.right)
        return root